File: EmojiPanel/src/main/java/com/binlee/emoji/span/EmojiSpan.java
package com.binlee.emoji.span;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.text.style.ImageSpan;

import androidx.annotation.NonNull;

public final class EmojiSpan extends ImageSpan {

    private int bgColor = Color.TRANSPARENT;

    public EmojiSpan(Drawable d, String source) {
        super(d, source);
    }

    @Override
    public void draw(@NonNull Canvas canvas, CharSequence text, int start, int end, float x, int top, int y, int bottom,
                     @NonNull Paint paint) {
        Paint.FontMetricsInt fm = paint.getFontMetricsInt();
        final Drawable d = getDrawable();
        canvas.save();
        // 计算y方向的位移
        int transY = (y + fm.descent + y + fm.ascent) / 2 - d.getBounds().bottom / 2;
        final int oldColor = paint.getColor();
        paint.setColor(bgColor);
        canvas.drawRect(x, top, x + d.getBounds().right, bottom, paint);
        // 绘制图片位移一段距离
        canvas.translate(x, transY);
        paint.setColor(oldColor);
        d.draw(canvas);
        canvas.restore();
    }

    @Override
    public int getSize(@NonNull Paint paint, CharSequence text, int start, int end, Paint.FontMetricsInt fm) {
        final Drawable d = getDrawable();
        final Rect rect = d.getBounds();
        if (fm != null) {
            final Paint.FontMetricsInt fontMetricsInt = paint.getFontMetricsInt();
            final int fontHeight = fontMetricsInt.bottom - fontMetricsInt.top;
            final int drawableHeight = rect.bottom - rect.top;
            final int top = drawableHeight / 2 - fontHeight / 4;
            final int bottom = drawableHeight / 2 + fontHeight / 4;
            fm.ascent = -bottom;
            fm.top = -bottom;
            fm.bottom = top;
            fm.descent = top;
        }
        return rect.right;
    }

    public void updateColor(int color) {
        bgColor = color;
        getDrawable().invalidateSelf();
    }
}

File: EmojiPanel/src/main/java/com/binlee/emoji/span/EmojiSpan.kt
package com.binlee.emoji.span

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Paint.FontMetricsInt
import android.graphics.drawable.Drawable
import android.text.style.ImageSpan

class EmojiSpan(d: Drawable?, source: String?) : ImageSpan(d!!, source!!) {

  private var bgColor = Color.TRANSPARENT

  override fun draw(
    canvas: Canvas, text: CharSequence, start: Int, end: Int, x: Float, top: Int, y: Int, bottom: Int,
    paint: Paint
  ) {
    val fm = paint.fontMetricsInt
    val d = drawable
    canvas.save()
    // 计算y方向的位移
    val transY = (y + fm.descent + y + fm.ascent) / 2 - d.bounds.bottom / 2
    val oldColor = paint.color
    paint.color = bgColor
    canvas.drawRect(x, top.toFloat(), x + d.bounds.right, bottom.toFloat(), paint)
    // 绘制图片位移一段距离
    canvas.translate(x, transY.toFloat())
    paint.color = oldColor
    d.draw(canvas)
    canvas.restore()
  }

  override fun getSize(paint: Paint, text: CharSequence, start: Int, end: Int, fm: FontMetricsInt?): Int {
    val d = drawable
    val rect = d.bounds
    if (fm != null) {
      val fontMetricsInt = paint.fontMetricsInt
      val fontHeight = fontMetricsInt.bottom - fontMetricsInt.top
      val drawableHeight = rect.bottom - rect.top
      val top = drawableHeight / 2 - fontHeight / 4
      val bottom = drawableHeight / 2 + fontHeight / 4
      fm.ascent = -bottom
      fm.top = -bottom
      fm.bottom = top
      fm.descent = top
    }
    return rect.right
  }

  fun updateColor(color: Int) {
    bgColor = color
    drawable.invalidateSelf()
  }
}